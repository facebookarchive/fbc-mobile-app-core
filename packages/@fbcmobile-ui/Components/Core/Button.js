/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

'use strict';

import type {ViewStyleProp} from 'react-native/Libraries/StyleSheet/StyleSheet';

import * as React from 'react';
import Text from '@fbcmobile/ui/Components/Core/Text';
import {Colors} from '@fbcmobile/ui/Theme';
import {StyleSheet, TouchableOpacity} from 'react-native';

type ButtonVariant = 'primary' | 'plain' | 'secondary' | 'text' | 'warning';
type TextAlign = 'left' | 'center';

type Props = {
  +children: React.Node,
  +variant?: ButtonVariant,
  +textAlign?: TextAlign,
  +fontSize?: number,
  +onPress?: void | (() => void | Promise<void>),
  +disabled?: boolean,
  +style?: ViewStyleProp,
  +rightIcon?: React.Node,
};

const Button = ({
  children,
  onPress,
  disabled = false,
  variant = 'primary',
  style,
  textAlign = 'center',
  fontSize = 16,
  rightIcon,
}: Props) => {
  return (
    <TouchableOpacity
      style={StyleSheet.compose(
        [
          styles.root,
          styles[`${variant}Variant`],
          disabled ? styles[`${variant}Disabled`] : null,
        ],
        style,
      )}
      onPress={onPress}
      disabled={disabled}>
      <Text
        variant="h3"
        weight="bold"
        style={[
          styles.text,
          styles[`${variant}Text`],
          {fontSize: fontSize},
          disabled
            ? variant === 'text'
              ? styles.disabledTextVariant
              : styles.disabledText
            : null,
          styles[`${textAlign}TextAlign`],
        ]}>
        {children}
      </Text>
      {rightIcon}
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  root: {
    paddingVertical: 10.5,
    paddingHorizontal: 12,
    borderRadius: 4,
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
  },
  primaryVariant: {
    backgroundColor: Colors.Blue,
  },
  primaryDisabled: {
    backgroundColor: Colors.DisabledBackground,
  },
  secondaryVariant: {
    backgroundColor: Colors.BlueLight,
  },
  secondaryDisabled: {
    backgroundColor: Colors.DisabledBackground,
  },
  plainVariant: {
    backgroundColor: Colors.White,
    borderWidth: 1,
    borderColor: Colors.BlueGray,
  },
  plainDisabled: {
    backgroundColor: Colors.White,
    borderColor: Colors.DisabledText,
  },
  warningVariant: {
    backgroundColor: Colors.RedBackground,
  },
  warningDisabled: {
    backgroundColor: Colors.DisabledBackground,
  },
  textText: {
    color: Colors.Blue,
  },
  disabledTextVariant: {
    color: Colors.BlueTransparent,
  },
  disabledText: {
    color: Colors.DisabledText,
  },
  primaryText: {
    color: Colors.White,
  },
  secondaryText: {
    color: Colors.Blue,
  },
  plainText: {
    color: Colors.BlueGray,
  },
  warningText: {
    color: Colors.BrightRed,
  },
  leftTextAlign: {
    textAlign: 'left',
  },
  centerTextAlign: {
    textAlign: 'center',
  },
  text: {
    display: 'flex',
    flexGrow: 1,
    flexShrink: 1,
  },
});

export default Button;
