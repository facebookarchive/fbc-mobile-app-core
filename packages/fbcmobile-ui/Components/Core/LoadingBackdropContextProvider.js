/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

'use strict';

import * as React from 'react';
import Spinner from 'react-native-spinkit';
import {Colors} from '@fbcmobile/ui/Theme';
import {Dimensions, StyleSheet, View} from 'react-native';

type LoadingBackdropContextType = {|
  setIsLoading: (isLoading: boolean) => void,
|};

export const LoadingBackdropContext = React.createContext<LoadingBackdropContextType>(
  {
    setIsLoading: () => {},
  },
);

type Props = {|
  +children: React.Node,
|};

const LoadingBackdropContextProvider = ({children}: Props) => {
  const [isLoading, setIsLoading] = React.useState(false);
  return (
    <LoadingBackdropContext.Provider value={{setIsLoading}}>
      {isLoading ? (
        <View style={styles.root}>
          <Spinner size={60} color={Colors.White} type="ThreeBounce" />
        </View>
      ) : null}
      {children}
    </LoadingBackdropContext.Provider>
  );
};

const styles = StyleSheet.create({
  root: {
    position: 'absolute',
    zIndex: 1,
    height: Dimensions.get('window').height,
    width: Dimensions.get('window').width,
    backgroundColor: Colors.TransparentBlack,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
});

export default LoadingBackdropContextProvider;
