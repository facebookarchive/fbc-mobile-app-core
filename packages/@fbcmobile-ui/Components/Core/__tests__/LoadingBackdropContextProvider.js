/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

'use strict';

import Button from '@fbcmobile/ui/Components/Core/Button';
import LoadingBackdropContextProvider from '../LoadingBackdropContextProvider';
import React from 'react';
import renderer from 'react-test-renderer';

describe('LoadingBackdropContextProvider Tests', () => {
  test('LoadingBackdropContextProvider renders correctly', () => {
    global.managedClock(() => {
      const tree = renderer
        .create(
          <LoadingBackdropContextProvider>
            <Button>Test</Button>
          </LoadingBackdropContextProvider>,
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
