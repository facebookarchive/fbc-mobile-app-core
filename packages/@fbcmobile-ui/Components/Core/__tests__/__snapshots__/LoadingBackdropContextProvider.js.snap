// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingBackdropContextProvider Tests LoadingBackdropContextProvider renders correctly 1`] = `
<View
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#1877F2",
      "borderRadius": 4,
      "display": "flex",
      "flexDirection": "row",
      "opacity": 1,
      "paddingHorizontal": 12,
      "paddingVertical": 10.5,
    }
  }
>
  <Text
    style={
      Array [
        Array [
          Object {
            "color": "#484848",
            "fontSize": 16,
            "fontWeight": "500",
            "lineHeight": 20,
          },
          null,
          Object {
            "fontFamily": "Roboto-Bold",
          },
          Object {
            "textAlign": "left",
          },
        ],
        Array [
          Object {
            "display": "flex",
            "flexGrow": 1,
            "flexShrink": 1,
          },
          Object {
            "color": "white",
          },
          Object {
            "fontSize": 16,
          },
          null,
          Object {
            "textAlign": "center",
          },
        ],
      ]
    }
  >
    Test
  </Text>
</View>
`;
